sweepbatcher/presigned.go

```diff
@@ -0,0 +1,457 @@
+	}
+
+	return presign(
+		ctx, b.cfg.presignedHelper, destAddr, sweeps, nextBlockFeeRate,
+	)
+}
+
+// presigner tries to presign a batch transaction.
+type presigner interface {
+	// Presign tries to presign a batch transaction. If the method returns
+	// nil, it is guaranteed that future calls to SignTx on this set of
+	// sweeps return valid signed transactions.
+	Presign(ctx context.Context, tx *wire.MsgTx,
+		inputAmt btcutil.Amount) error
+}
+
+// presign tries to presign batch sweep transactions of the sweeps. It signs
+// multiple versions of the transaction to make sure there is a transaction to
+// be published if minRelayFee grows. If feerate is high, then a presigned tx
+// gets LockTime equal to timeout minus 50 blocks, as a precautionary measure.
+// A feerate is considered high if it is at least 100 sat/vbyte AND is at least
+// 10x of the current next block feerate.
+func presign(ctx context.Context, presigner presigner, destAddr btcutil.Address,
+	sweeps []sweep, nextBlockFeeRate chainfee.SatPerKWeight) error {
+
+	if presigner == nil {
+		return fmt.Errorf("presigner is not installed")
+	}
+
+	if len(sweeps) == 0 {
+		return fmt.Errorf("there are no sweeps")
+	}
+
+	if nextBlockFeeRate == 0 {
+		return fmt.Errorf("nextBlockFeeRate is not set")
+	}
+
+	// Keep track of the total amount this batch is sweeping back.
+	batchAmt := btcutil.Amount(0)
+	for _, sweep := range sweeps {
+		batchAmt += sweep.value
+	}
+
+	// Find the sweep with the earliest expiry.
+	timeout := sweeps[0].timeout
+	for _, sweep := range sweeps[1:] {
+		timeout = min(timeout, sweep.timeout)
+	}
+	if timeout <= 0 {
+		return fmt.Errorf("timeout is invalid: %d", timeout)
+	}
+
+	// Go from the floor (1.01 sat/vbyte) to 2k sat/vbyte with step of 1.2x.
+	const (
+		start            = chainfee.FeePerKwFloor
+		stop             = chainfee.AbsoluteFeePerKwFloor * 2_000
+		factorPPM        = 1_200_000
+		timeoutThreshold = 50
+	)
+
+	// Calculate the locktime value to use for high feerate transactions.
+	highFeeRateLockTime := uint32(timeout - timeoutThreshold)
+
+	// Calculate which feerate to consider high. At least 100 sat/vbyte and
+	// at least 10x of current nextBlockFeeRate.
+	highFeeRate := max(100*chainfee.FeePerKwFloor, 10*nextBlockFeeRate)
+
+	// Set LockTime to 0. It is not critical.
+	const currentHeight = 0
+
+	for fr := start; fr <= stop; fr = (fr * factorPPM) / 1_000_000 {
+		// Construct an unsigned transaction for this fee rate.
+		tx, _, feeForWeight, fee, err := constructUnsignedTx(
+			sweeps, destAddr, currentHeight, fr,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to construct unsigned tx "+
+				"for feeRate %v: %w", fr, err)
+		}
+
+		// If the feerate is high enough, set locktime to prevent
+		// broadcasting such a transaction too early by mistake.
+		if fr >= highFeeRate {
+			tx.LockTime = highFeeRateLockTime
+		}
+
+		// Try to presign this transaction.
+		err = presigner.Presign(ctx, tx, batchAmt)
```

Comment:
Just a side question: Currently the result is not yet used. Is it intended to then pass this pre-signed transaction into our store? I guess once the batch is confirmed we could clean up all the presigned fee versions.

Reply:
`PresignHelper` is responsible for storing the transactions. SweepBatcher relies on `helper.SignTx` to be able to load transactions stored by previous `helper.Presign` and `helper.SignTx` calls.

> I guess once the batch is confirmed we could clean up all the presigned fee versions.

This is done. There is the method `CleanupTransactions` which is called by `batch.handleConf` method.